contracts/Pool.sol

```diff
@@ -14,260 +15,368 @@ import "./IBorrowersRegistry.sol";
+    require(totalSupply() - totalBorrowed() >= _amount, "There is no enough deposit in the pool to fund the loan.");
 
-        emit Repayment(msg.sender, msg.value, block.timestamp);
-    }
+    _accumulateBorrowingInterests(msg.sender);
 
+    borrowed[msg.sender] += _amount;
+    borrowed[address(this)]+= _amount;
 
-    /* ========== VIEW FUNCTIONS ========== */
+    payable(msg.sender).transfer(_amount);
 
+    _updateRates();
 
-    /**
-     * Returns the current borrowed amount for the given user
-     * The value includes the interest rates owned at the current moment
-     * @dev _user the address of queried borrower
-    **/
-    function getBorrowed(address _user) public view returns(uint256) {
-        return borrowIndex.getIndexedValue(borrowed[_user], _user);
-    }
+    emit Borrowing(msg.sender, _amount, block.timestamp);
+  }
 
+  /**
+   * Repays the message value
+   * It updates user borrowed balance, total borrowed amount and rates
+  **/
+  function repay() payable external {
+    _accumulateBorrowingInterests(msg.sender);
 
-    /**
-     * Returns the current deposited amount for the given user
-     * The value includes the interest rates earned at the current moment
-     * @dev _user the address of queried depositor
-    **/
-    function getDeposits(address user) public view returns(uint256) {
-      return depositIndex.getIndexedValue(deposits[user], user);
-    }
+    require(borrowed[address(this)] >= msg.value, "You are trying to repay more that was borrowed by all users.");
+    require(borrowed[msg.sender] >= msg.value, "You are trying to repay more that was borrowed by user.");
 
+    borrowed[msg.sender] -= msg.value;
+    borrowed[address(this)] -= msg.value;
 
-    /**
-     * Returns the current interest rate for deposits
-    **/
-    function getDepositRate() public view returns(uint256) {
-      return ratesCalculator.calculateDepositRate(totalBorrowed, totalDeposited);
-    }
+    _updateRates();
 
+    emit Repayment(msg.sender, msg.value, block.timestamp);
+  }
 
-    /**
-     * Returns the current interest rate for borrowings
-    **/
-    function getBorrowingRate() public view returns(uint256) {
-      return ratesCalculator.calculateBorrowingRate(totalBorrowed, totalDeposited);
-    }
 
+  /* =========
 
-    /* ========== INTERNAL FUNCTIONS ========== */
 
+  /**
+   * Returns the current borrowed amount for the given user
+   * The value includes the interest rates owned at the current moment
+   * @dev _user the address of queried borrower
+  **/
+  function getBorrowed(address _user) public view returns (uint256) {
+    return borrowIndex.getIndexedValue(borrowed[_user], _user);
+  }
 
-    function updateRates() internal {
-      depositIndex.setRate(ratesCalculator.calculateDepositRate(totalBorrowed, totalDeposited));
-      borrowIndex.setRate(ratesCalculator.calculateBorrowingRate(totalBorrowed, totalDeposited));
-    }
+  function totalSupply() public view override returns (uint256) {
```

Comment:
We should index this value in a similar way to balanceOf().

We should add test cases checking if the sum of  balanceOf of every user is equal to totalSupply in different scenarios

Reply:
Could you elaborate on your first point, please? I am not sure if I got your point as all that `totalSupply()` does is calling the mentioned `balanceOf()` function.